test/test.h

```diff
@@ -1,49 +1,116 @@
-	int (*test_fnc)(void);
-	/// @a test_fnc is executed with that many cores
-	unsigned threads_count;
-	/// The command line arguments passed to the wrapped main function
-	const char **test_arguments;
-};
-
-/// The test configuration object, must be defined by the test sources
-extern const struct test_config test_config;
-extern bool test_thread_barrier(void);
-
-// core.h typedefs
-typedef uint64_t lp_id_t;
-typedef unsigned rid_t;
-typedef int nid_t;
-
-extern rid_t n_threads;
-extern __thread rid_t rid;
+static struct {
+	jmp_buf fail_buffer;
+	int ret;
+	unsigned total;
+	unsigned passed;
+	unsigned failed;
+	unsigned xfailed;
+	unsigned uxpassed;
+	unsigned should_pass;
+	unsigned should_fail;
+} test_unit;
+
+
+#define assert(condition) do {                                           \
+        if(!(condition)) {                                               \
+                fprintf(stderr, "assertion failed: " #condition " at %s:%d\n", __FILE__, __LINE__);   \
+                test_unit.ret = -1;                                       \
+        }                                                                \
+} while(0)
+
+#define init() do {                              \
+        if(setjmp(test_unit.fail_buffer)) {      \
+                test_unit.ret = -1;              \
+                finish();                        \
+        }                                        \
+} while(0)
+
+#define check_passed_asserts() do {      \
+	int ret = test_unit.ret; \
+	test_unit.ret = 0;       \
+	return ret;              \
+} while(0)
+
+#define finish() do { \
+        int d1 = snprintf(NULL, 0, "PASSED.............: %u / %u\n", test_unit.passed, test_unit.should_pass); \
+        int d2 = snprintf(NULL, 0, "EXPECTED FAIL......: %u / %u\n", test_unit.xfailed, test_unit.should_fail);\
+        int d3 = snprintf(NULL, 0, "FAILED.............: %u\n", test_unit.failed);                             \
+        int d4 = snprintf(NULL, 0, "UNEXPECTED PASSED..: %u\n", test_unit.uxpassed);                           \
+        int d = ((d1 > d2 && d1 > d3 && d1 > d4) ? d1: ((d2 > d3 && d2 > d4) ? d2 : (d3 > d4 ? d3 : d4)));     \
+        printf("%.*s\n", d, "============================================================================");   \
+        printf("PASSED.............: %u / %u\n", test_unit.passed, test_unit.should_pass);                     \
+	printf("EXPECTED FAIL......: %u / %u\n", test_unit.xfailed, test_unit.should_fail);                    \
+        printf("FAILED.............: %u\n", test_unit.failed);                                                 \
+	printf("UNEXPECTED PASSED..: %u\n", test_unit.uxpassed);                                               \
+	printf("%.*s\n", d, "============================================================================");   \
+	return test_unit.ret;                                                                                  \
```

Comment:
It looks neat on the outside, but a return statement in a macro can be confusing and hard to understand, especially when trying to debug a failing test

Reply:
Same as the previous comment: if there is a structure in the testing framework, I see no problem. It's a support to development, which is intended to be used by us with some rules. We can discuss it anyhow.

Reply:
This no longer applies as it has been converted into a function.